Feature/chart, enhanced websocket support, dynamic generating horizontaltablist

gmverdon opened the pull request.

gmverdon committed:
Add TradingView chart

gmverdon committed:
TradingView widget module and cleanup

gmverdon committed:
Generate horizontalList from props

gmverdon committed:
Update tradingview widget module

gmverdon committed:
Shouldcomponent update and generate tradingview from props

gmverdon committed:
Update charting styling

gmverdon committed:
Price websockets change based on selected crypto, get full cryptolist

gmverdon committed:
Get currentprice when selecting new crypto

gmverdon committed:
Update defaults in state

rafaelklaessen commented on src/components/Chart/index.js:
```diff
@@ -0,0 +1,27 @@
+
+class Chart extends Component {
+
+  shouldComponentUpdate = (nextProps) => this.props.selectedCrypto !== nextProps.selectedCrypto
```
Is this actually necessary? TradingViewWidget does this too.

gmverdon replied:
Removed

gmverdon commented on src/scenes/TradeHub/index.js:
```diff
@@ -71,11 +114,9 @@ class TradeHub extends Component {
   }
 
   setHighestPrice(currentPrice) {
```
name parameter to price

gmverdon committed:
Remove redundant shouldComponentUpdate

TradingViewWidget is a pureComponent which already handles redundant rerenders.

gmverdon committed:
Change price paramaters to more general name

rafaelklaessen commented on src/components/Chart/index.js:
```diff
@@ -0,0 +1,24 @@
+
+class Chart extends Component {
+  render() {
+    let symbol = "BINANCE:" + this.props.selectedCrypto.symbol;
```
Use `const` instead of `let`

rafaelklaessen commented on src/components/HorizontalTabList/index.js:
```diff
@@ -1,89 +1,32 @@
 class HorizontalTabList extends Component {
+
   getListItems() {
+    let list = this.props.list;
```
Use `const` instead of `let`

rafaelklaessen commented on src/components/HorizontalTabList/index.js:
```diff
@@ -1,89 +1,32 @@
-    for (var i = 0; i < tilesData.length; i++) {
-      let newItem = this.generateListItem(tilesData[i], i)
+    for (var i = 0; i < list.length; i++) {
+      let newItem = this.generateListItem(list[i].symbol, i)
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+
+  bindSocket = (symbol) => {
+    binance.websockets.candlesticks([symbol], "1m", (candlesticks) => {
+      let { k:ticks } = candlesticks;
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+  bindSocket = (symbol) => {
+    binance.websockets.candlesticks([symbol], "1m", (candlesticks) => {
+      let { k:ticks } = candlesticks;
+      let { c:close } = ticks;
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+      let { k:ticks } = candlesticks;
+      let { c:close } = ticks;
       //console.log(symbol+" "+interval+" candlestick update");
       let currentPrice = parseFloat(close);
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
-
-      self.setState({
+      this.setState({
         currentPrice: currentPrice
```
Use
```javascript
this.setState({ currentPrice });
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then((res) => {
+      return res.json();
+    }).then((data) => {
+      let cryptoList = [];
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then((res) => {
+      return res.json();
```
```javascript
.then(res => res.json())
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+    }).then((data) => {
+      let cryptoList = [];
+
+      let symbols = data.symbols;
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+
+      if (cryptoList.length > 0) {
+        this.setState({
+          cryptoList: cryptoList,
```
Remove `: cryptoList` as it's redundant

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+      let cryptoList = [];
+
+      let symbols = data.symbols;
+      for (var i = 0; i < symbols.length; i++) {
```
Use `let` instead of `var`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
-  isHighestPrice(currentPrice) {
-    let isHighestPrice = currentPrice > this.state.highestPrice;
+  isHighestPrice(price) {
+    let isHighestPrice = price > this.state.highestPrice;
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,126 @@ import binance from 'binance-api';
+  shouldSell(price) {
     if (this.state.sellEnabled && !this.state.hasSold) {
-      let shouldSell = (currentPrice <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
+      let shouldSell = (price <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
   }
 
+  changeSelectedCrypto(symbol) {
+    let crypto = this.state.cryptoList.find((obj) => { return obj.symbol === symbol; });
```
Use `const` instead of `let`

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
 
+  changeSelectedCrypto(symbol) {
+    let crypto = this.state.cryptoList.find((obj) => { return obj.symbol === symbol; });
+    if (crypto == null) { return; }
```
Use
```javascript
if (crypto === null) return;
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
+
+    this.setState({
+      selectedCrypto: crypto
+    },() => { this.rebindSocket(); });
```
() => this.rebindSocket(); is enough

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -112,22 +152,44 @@ class TradeHub extends Component {
+    let currentPrice = this.state.currentPrice.toFixed(6);
+    let highestPrice = this.state.highestPrice.toFixed(6);;
+    let sellPrice = (highestPrice - highestPrice * this.state.diffPercentage).toFixed(6);
+    let cryptoList = this.state.cryptoList;
```
Use `const` instead of `let`

gmverdon committed:
Code cleanup and change default settings

gmverdon committed:
Set highestPrice to currentPrice when selecting new cryptocurrency

gmverdon committed:
Move redundant cryptoList key in setState

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
+  isHighestPrice(price) {
+    const isHighestPrice = price > this.state.highestPrice;
     return isHighestPrice;
   }
```
Also possible:
```javascript
isHighestPrice = price => price > this.state.highestPrice;
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
+      const shouldSell = (price <= (this.state.highestPrice - this.state.highestPrice * this.state.diffPercentage));
       return shouldSell;
     }
   }
```
How about using object extraction here?
```javascript
const { sellEnabled, hasSold, highestPrice, diffPercentage } = this.state;
if (sellEnabled && !hasSold) {
  return price <= highestPrice - highestPrice * diffPercentage;
}
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
+    alert("SOLD at " + price);
     this.setState({
       hasSold: true
     })
```
Add ;

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -100,7 +134,7 @@ class TradeHub extends Component {
-    let diffPercentage = (e.target.value / 100);
+    const diffPercentage = (e.target.value / 100);
     this.setState({
       diffPercentage: diffPercentage
```
Can be written as
```javascript
this.setState({ diffPercentage });
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -112,22 +146,46 @@ class TradeHub extends Component {
+    binance.prices((ticker) => {
+      const currentPrice = parseFloat(ticker[crypto.symbol]);
+      this.setState({
+        currentPrice: currentPrice,
```
`currentPrice` twice is redundant

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -112,22 +146,46 @@ class TradeHub extends Component {
   }
 
+  changeSelectedCrypto(symbol) {
+    const crypto = this.state.cryptoList.find((obj) => { return obj.symbol === symbol; });
```
Use
```javascript
this.state.cryptoList.find(obj => obj.symbol === symbol);
```

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
+      const cryptoList = [];
+      for (let i = 0; i < data.symbols.length; i++) {
+          cryptoList.push(data.symbols[i]);
+      }
```
Use
```javascript
const cryptoList = data.symbols;
```

gmverdon replied:
Nice one!

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
-      if (this.isHighestPrice(currentPrice)) {
-        this.setHighestPrice(currentPrice);
+  getCryptoList() {
+    fetch("https://api.binance.com/api/v1/exchangeInfo").then(res => res.json()).then((data) => {
```
Use single quotes

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,86 +8,120 @@ import binance from 'binance-api';
-        this.sell(currentPrice);
+      if (this.shouldSell(price)) {
+        this.sell(price);
         return;
```
Unnecessary return statement

gmverdon committed:
Cleanup code

gmverdon committed:
Cleanup shoudSell method

gmverdon committed:
Semicolumns and other code improvements

rafaelklaessen commented on src/components/Chart/index.js:
```diff
@@ -0,0 +1,24 @@
+// libraries
+import TradingViewWidget from 'react-tradingview-widget';
+
+class Chart extends Component {
```
Use a functional component?

rafaelklaessen commented on src/components/Chart/index.js:
```diff
@@ -0,0 +1,24 @@
+import React, { Component } from 'react';
+import './styles.css';
+
+// Bootstrap components
```
That's obvious

rafaelklaessen commented on src/scenes/TradeHub/index.js:
```diff
@@ -8,126 +8,178 @@ import binance from 'binance-api';
       }
+
+      if (this.shouldSell(price)) {
+        this.sell(price);
```
You *could* do
```javascript
if (this.shouldSell(price)) this.sell(price):
```
Without the curly braces

gmverdon committed:
Change to functional component

gmverdon committed:
InfoPanel to functional component

gmverdon committed:
Remove unneccesary import for Chart component
